src/propulsion/calculate_RPM.cpp

```diff
@@ -0,0 +1,142 @@
+
+CalculateRPM::CalculateRPM(Logger &log) : data_(data::Data::getInstance()),
+                                          acceleration_slip_(),
+                                          log_(log)
+{
+    state_ = data_.getStateMachineData();
+    motor_velocity_ = {0, 0, 0, 0};
+}
+
+std::vector<std::vector<double>> CalculateRPM::transpose(std::vector<std::vector<double>> data)
+{
+    std::vector<std::vector<double>> transpose(2, std::vector<double>(data.size(), 1));
+    for (uint16_t i = 0; i < data.size(); ++i) {
+        for (uint16_t j = 0; j < 2; ++j) {
+            transpose[j][i] = data[i][j];
+        }
+    }
+    return transpose;
+}
+
+bool CalculateRPM::check_file(std::string filepath)
+{
+    std::ifstream data;
+    data.open(filepath);
+
+    if (!data.is_open()) {
+        log_.ERR("MOTOR", "Could not open file: %s", filepath.c_str());
+        return false;
+    } else if (filepath == kAccelerationData) {
+        log_.INFO("MOTOR", "Calculating acceleration slip...");
+    } else {
+        log_.INFO("MOTOR", "Calculating deceleration slip...");
+    }
+
+    return true;
+};
+
+void CalculateRPM::RPM_calc(std::string filepath)
+{
+    double slip, translational_velocity, angular_velocity, rpm;
+    std::ifstream data;
+    data.open(filepath);
+
+    std::string line;
+    while (std::getline(data, line)) {
+        std::string split_line;
+        std::vector<double> temp_vec;
+        std::stringstream ss(line);
+        while (std::getline(ss, split_line, '\t')) {
+            temp_vec.push_back(std::move(stod(split_line)));
+        }
+
+        // Calculate angular velocity from slip, translational velocity and halbach wheel radius
+        slip = temp_vec[0];
+        translational_velocity = temp_vec[1];
+        angular_velocity = (slip + translational_velocity) / kHalbachRadius;
+
+        // Calculate RPM given calculated angular velocity
+        rpm = (angular_velocity * 60) / (2 * M_PI);
+
+        // Use temporary vector to hold translational velocity and rpm
+        temp_vec[0] = translational_velocity;
+        temp_vec[1] = rpm;
+
+        // Add data to the acceleration_slip_ queue
+        acceleration_slip_.push(temp_vec);
+    }
+
+    // // If both containers have been populated, set bool to true
+    // if (!acceleration_slip_.empty() && !deceleration_slip_.empty()) {
+    //     acceleration_slip_ = transpose(acceleration_slip_);
+    //     deceleration_slip_ = transpose(deceleration_slip_);
+    //     slip_calculated_ = true;
+    //     log_.INFO("MOTOR", "All slip values calculated");
+    // }
+};
+
+int32_t CalculateRPM::calculateRPM(float velocity)
+{
+    do {
+        if (velocity < acceleration_slip_.front()[0]) {
+            return acceleration_slip_.front()[1];
+        }
+        if (acceleration_slip_.size() == 1) {
+            return acceleration_slip_.front()[1];
+        }
+        acceleration_slip_.pop();
+    } while (acceleration_slip_.front()[0] < velocity);
```

Comment:
Make a check for when the queue is empty. Then return a fixed RPM 6000 for example

Reply:
Thanks for reminding. Fixed it.